File: java/functions/hello-auditlog/src/main/java/functions/Function.java
/*
 * Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package functions;

import com.google.cloud.functions.CloudEventsFunction;
import com.google.events.cloud.audit.v1.LogEntryData;
import com.google.protobuf.InvalidProtocolBufferException;
import com.google.protobuf.util.JsonFormat;
import io.cloudevents.CloudEvent;
import java.util.logging.Logger;

public class Function implements CloudEventsFunction {

  private static final Logger logger = Logger.getLogger(Function.class.getName());

  @Override
  public void accept(CloudEvent cloudEvent) throws InvalidProtocolBufferException {

    // CloudEvent information
    logger.info("Id: " + cloudEvent.getId());
    logger.info("Source: " + cloudEvent.getSource());
    logger.info("Type: " + cloudEvent.getType());

    String json = new String(cloudEvent.getData().toBytes());
    LogEntryData.Builder builder = LogEntryData.newBuilder();
    JsonFormat.parser().merge(json, builder);
    LogEntryData data = builder.build();

    logger.info("ProtoPayload: " + data.getProtoPayload());
    logger.info("ServiceName: " + data.getProtoPayload().getServiceName());
    logger.info("MethodName: " + data.getProtoPayload().getMethodName());
    logger.info("ResourceName: " + data.getProtoPayload().getResourceName());
  }
}

File: dotnet/functions/hello-auditlog/Function.cs
// Copyright 2022 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
using CloudNative.CloudEvents;
using Google.Cloud.Functions.Framework;
using Google.Events.Protobuf.Cloud.Audit.V1;
using System;
using System.Threading;
using System.Threading.Tasks;

namespace HelloAuditLog;

public class Function : ICloudEventFunction<LogEntryData>
{
    public Task HandleAsync(CloudEvent cloudEvent, LogEntryData data, CancellationToken cancellationToken)
    {
        Console.WriteLine("LogEntry information:");
        Console.WriteLine($"  ProtoPayload: {data.ProtoPayload}");
        Console.WriteLine($"    ServiceName: {data.ProtoPayload.ServiceName}");
        Console.WriteLine($"    MethodName: {data.ProtoPayload.MethodName}");
        Console.WriteLine($"    ResourceName: {data.ProtoPayload.ResourceName}");
        Console.WriteLine("CloudEvent information:");
        Console.WriteLine($"  ID: {cloudEvent.Id}");
        Console.WriteLine($"  Source: {cloudEvent.Source}");
        Console.WriteLine($"  Type: {cloudEvent.Type}");
        Console.WriteLine($"  Subject: {cloudEvent.Subject}");
        Console.WriteLine($"  DataSchema: {cloudEvent.DataSchema}");
        Console.WriteLine($"  DataContentType: {cloudEvent.DataContentType}");
        Console.WriteLine($"  Time: {cloudEvent.Time?.ToUniversalTime():yyyy-MM-dd'T'HH:mm:ss.fff'Z'}");
        Console.WriteLine($"  SpecVersion: {cloudEvent.SpecVersion}");

        return Task.CompletedTask;
    }
}